src/Libraries/Microsoft.Extensions.AI.Abstractions/Embeddings/BinaryEmbedding.cs

```diff
@@ -0,0 +1,123 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System;
+using System.Buffers;
+using System.Collections;
+using System.ComponentModel;
+using System.Text.Json;
+using System.Text.Json.Serialization;
+using Microsoft.Shared.Diagnostics;
+
+namespace Microsoft.Extensions.AI;
+
+/// <summary>Represents an embedding composed of a bit vector.</summary>
+public sealed class BinaryEmbedding : Embedding
+{
+    /// <summary>The embedding vector this embedding represents.</summary>
+    private BitArray _vector;
+
+    /// <summary>Initializes a new instance of the <see cref="BinaryEmbedding"/> class with the embedding vector.</summary>
+    /// <param name="vector">The embedding vector this embedding represents.</param>
+    /// <exception cref="ArgumentNullException"><paramref name="vector"/> is <see langword="null"/>.</exception>
+    public BinaryEmbedding(BitArray vector)
+    {
+        _vector = Throw.IfNull(vector);
+    }
+
+    /// <summary>Gets or sets the embedding vector this embedding represents.</summary>
+    [JsonConverter(typeof(VectorConverter))]
+    public BitArray Vector
+    {
+        get => _vector;
+        set => _vector = Throw.IfNull(value);
+    }
+
+    /// <inheritdoc />
+    [JsonIgnore]
+    public override int Dimensions => _vector.Length;
+
+    /// <summary>Provides a <see cref="JsonConverter{BitArray}"/> for serializing <see cref="BitArray"/> instances.</summary>
+    [EditorBrowsable(EditorBrowsableState.Never)]
+    public sealed class VectorConverter : JsonConverter<BitArray>
+    {
+        /// <inheritdoc/>
+        public override BitArray Read(ref Utf8JsonReader reader, Type typeToConvert, JsonSerializerOptions options)
+        {
+            _ = Throw.IfNull(typeToConvert);
+            _ = Throw.IfNull(options);
+
+            if (reader.TokenType != JsonTokenType.String)
+            {
+                throw new JsonException("Expected string property.");
+            }
+
+            ReadOnlySpan<byte> utf8;
```

Comment:
Nit: much of this branching can be simplified by writing
```C#
int length = reader.HasValueSequence ? checked((int)reader.ValueSequence.Length) : reader.ValueSpan.Length;
```
In practice, it's vanishingly likely for the underlying sequence to be larger than `int.MaxValue`. What's more, `JsonSerializer` never uses `ReadOnlySpan<byte>` readers and always goes with spans.

Reply:
Are you saying I shouldn't validate the sequence length? The structure is as it is in order to avoid accessing ValueSequence twice.

Reply:
You shouldn't need to access `ValueSequence` twice, just get the length to allocate an appropriate buffer and then pass that to `CopyString`.

Reply:
If the length overflows, then it's relying on either ArrayPool renting or CopyString failing, which feels weak. Can that not happen for some reason?

Reply:
> while avoiding the duplicated code in the Span and Sequence cases.

The duplicated code being the Rent call?